server/upsd.c

```diff
@@ -303,6 +303,17 @@ static void setuptcp(stype_t *server)
 			fatal_with_errno(EXIT_FAILURE, "setuptcp: setsockopt");
 		}
 
+		/* Ordinarily we request that IPv6 listeners handle only IPv6.
+		 * TOTHINK: Does any platform need `#ifdef IPV6_V6ONLY` given
+		 * that we apparently already have AF_INET6 OS support everywhere?
+		 */
+		if (ai->ai_family == AF_INET6) {
+			if (setsockopt(sock_fd, IPPROTO_IPV6, IPV6_V6ONLY, (void *)&one, sizeof(one)) != 0) {
+				upsdebug_with_errno(3, "setuptcp: setsockopt IPV6_V6ONLY");
```

Comment:
It's an interesting question if this approach is reasonable.

On one hand, I view it as a bug that systems default to using mapped addresses.   It's clearly a bug if programs rely on this and don't work well when v6only=0.  But once a program has code to handle multiple sockets (rather than opening just one), it seems cleaner to have one socket per AF.   So one can take the view that the system having v6only=0 is an accomodation for programs that haven't yet been fixed, and that programs that are fixed should set v6only=1.

On the other hand, if a system makes a choice, and the sysadmin doesn't change it, it seems best to try to respect that choice, if at all reasonable.  Given the rationale above, I find this an extremely weak argument.

Finally, I live in a world where v6only=1, and I think that's where the rest of the world ought to be.   So this code to me is just using a few cycles at startup (which is fine) and not changing anything, and it doesn't bother me personally.

All in all, I've convinced myself that setting v6only=1 on the socket is a fine thing to do.

Reply:
As for "a bug", Sam pointed to https://lwn.net/Articles/688462/ discussion from 2016 about the difference in opinions - both summarized in the initial post and flared up in replies.

The gist of it is that for best or worse, the RFC recommendation is to use mapped addresses - so most OSes do it deliberately to uphold the standard and thus make it into a sort of common law, and most others do not rock the boat so as to not lose relevance as "strangely incompatible" (some have it as a configurable toggle for default socket settings, with default being for mapped addresses per RFC).

On the practical side, it is more up to application software programs (like NUT) to know and care about the difference, and in fact those that do care about it (and get "fixed") - do so in spite of the standard :)

Reply:
The RFC when written may have recommended that, but I don't see that this means it is still consensus.   But we don't have to resolve that, if we agree that it is reasonable for individual programs that have explicit support for opening a socket per AF can just set v6only=1 for the socket.

Reply:
Going through the various stages, I did first want to rebuff the suggestions ("oh I've just got it to work", "ain't broke don't fix", etc. style of blabber) and cycled to "acceptance" - that indeed we should aim to set up separate listening sockets per address family here too, and only end up with one (though not crash) if the OS is deficient in some way. Tinkering on that now... so thanks! ;)